### DISCUSSION TITLE
Execute Batch Failling with unresolve token

### DISCUSSION THREAD

**Discussion by nepgya2803** [pnp/pnpcore#1359] [Q&A]
We are deploying the code to the customer environment. It works normally on our dev environment and the Sharepoint list configuration of us is the same. On our dev environment, we can add 6000 items perfectly while it only could add 2020 items (1000 items per batch) on the customer environment. Since the code is hosted and compiled on the AWS lambda function, we cannot debug directly, the only log returned from the customer environment is:

> 7ee91d6f-7709-4877-881e-0cb6a7e00a66	fail	PnP.Core.ClientError
PnP.Core.ClientException: Unresolved tokens found in API call https://(domain).sharepoint.com/sites/yh-dev-test/_api/web/lists/getbyid(guid'68458ee1-2f5d-49bd-a555-508830d75ed1')/items({Id})/ValidateUpdateListItem
   at PnP.Core.Services.BatchClient.CheckForUnresolvedTokens(Batch batch)
   at PnP.Core.Services.BatchClient.ExecuteBatch(Batch batch)
   at PnP.Core.Services.PnPContext.ExecuteAsync(Batch batch, Boolean throwOnError)
   at S3SPSync.StaffService.SPDataProcessor.ProcessDataAsync(List`1 Data)

```csharp
    var myList = await _context.Web.Lists.GetByTitleAsync(_input.StaffListName,
                                         p => p.Title,
                                         p => p.Items,
                                         p => p.Fields.QueryProperties(
                                           p => p.InternalName,
                                           p => p.FieldTypeKind,
                                           p => p.TypeAsString,
                                           p => p.Title));

    // Get all the user in site
    _userList = _context.Web.SiteUsers.AsRequested();
    CreateFieldDictiontary(myList);

    int addCount = 0;
    int updateCount = 0;
    string viewXml = CAMLQueryGenerator();
    var listExistingItem = await GetWithCalmQuery(myList, viewXml);
    Console.WriteLine($"There are {listExistingItem.Count()} in the current {_input.StaffListName} list");

    var addBatch = _context.NewBatch();
    var updateBatch = _context.NewBatch();

    foreach (var user in Data)
    {
            // Execute the upsert procedure
            if (!listExistingItem.Any()
                || !listExistingItem.Where(p => p["EmployeeNumber"].ToString() == user.personIdExternal).Any())
            {
                if (addCount == 999)
                {
                        await this._context.ExecuteAsync(addBatch);
                        addBatch = _context.NewBatch();
                        addCount = 0;
                    }
                }

                await Add(this._context, myList, user, addBatch);
                addCount++;
            }
            else
            {
                if (updateCount == 999)
                {
                        await this._context.ExecuteAsync(updateBatch);
                        updateBatch = _context.NewBatch();
                        updateCount = 0;             
                }

                await Update(this._context, myList, user, updateBatch);
                updateCount++;
            }
        }
    await this._context.ExecuteAsync(addBatch);
    await this._context.ExecuteAsync(updateBatch);
```

**Comment by jansenbe** [comment]
@ngqtrieu99 : why are you trying to group requests in batches like that? PnP Core SDK will automatically use multiple calls under the covers whenever a batch is too large (see https://pnp.github.io/pnpcore/using-the-sdk/basics-batching.html#batch-limits). 

Without doing a detailed check of your code I'm suspecting that somehow you are trying to update and item that's being added in the same batch or that was not loaded again after having been added.

**Reply by nepgya2803** [reply]
I also think of that but I'm sure that in the source data, every item with an Employee Number is unique so they cannot update an item being added.

**Reply by jansenbe** [reply]
@ngqtrieu99 : without reproduction steps it's hard to check if this is a bug and if so, how to solve it
